be/src/olap/aggregate_func.h

```diff
@@ -15,221 +15,283 @@
+    // saved in src.
+    //
+    // Memory Note: All heap memory allocated in init and update function should be freed
+    // before this function return. Memory allocated from arena will be still available
+    // and will be freed by client.
+    inline void finalize(void* src, Arena* arena) const {
+        _finalize_fn((char*)src, arena);
+    }
 
-template<FieldType field_type>
-struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_NONE, field_type> {
-    static void aggregate(char* left, const char* right, Arena* arena) {}
+private:
+    void (*_init_fn)(char* dst, Arena* arena);
+    void (*_update_fn)(char* dst, const char* src, Arena* arena);
+    void (*_merge_fn)(char* dst, const char* src, Arena* arena);
+    void (*_finalize_fn)(char* dst, Arena* arena);
+
+    friend class AggregateFuncResolver;
+
+    template<typename Traits>
+    AggregateInfo(const Traits& traits);
+};
+
+struct BaseAggregateFuncs {
+    // Default init function will set to null
+    static void init(char* dst, Arena* arena) {
+        *reinterpret_cast<bool*>(dst) = true;
+    }
+
+    // Default update do nothing.
+    static void update(char* dst, const char* src, Arena* arena) {
+    }
+
+    // For most aggregate method, its merge and update are same. If merge
+    // is same with update, keep merge nullptr to avoid duplicate code.
+    // AggregateInfo constructor will set merge function to update function
+    // when merge is nullptr.
+    AggUpdateFunc merge = nullptr;
+
+    // Default finalize do nothing.
+    static void finalize(char* src, Arena* arena) {
+    }
+};
+
+template<FieldAggregationMethod agg_method, FieldType field_type>
+struct AggregateFuncTraits : public BaseAggregateFuncs {
 };
 
 template <FieldType field_type>
-struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_MIN, field_type> {
-    static void aggregate(char* left, const char* right, Arena* arena) {
-        typedef typename FieldTypeTraits<field_type>::CppType CppType;
-        bool l_null = *reinterpret_cast<bool*>(left);
-        bool r_null = *reinterpret_cast<const bool*>(right);
-        if (l_null) {
-            return;
-        } else if (r_null) {
-            *reinterpret_cast<bool*>(left) = true;
-        } else {
-            CppType* l_val = reinterpret_cast<CppType*>(left + 1);
-            const CppType* r_val = reinterpret_cast<const CppType*>(right + 1);
-            if (*r_val < *l_val) { *l_val = *r_val; }
+struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_MIN, field_type> : public BaseAggregateFuncs {
+    typedef typename FieldTypeTraits<field_type>::CppType CppType;
+
+    static void update(char* dst, const char* src, Arena* arena) {
+        bool src_null = *reinterpret_cast<const bool*>(src);
+        // ignore null value
+        if (src_null) return;
+
+        bool dst_null = *reinterpret_cast<bool*>(dst);
+        CppType* dst_val = reinterpret_cast<CppType*>(dst + 1);
```

Comment:
if dst_null == true, dst + 1 has non-sense, it will be undefined behavior.

Reply:
OK, I will fix it.

Reply:
OK, I will fix it.